src/lib/storage/table.cpp

```diff
@@ -78,13 +93,62 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
 
 const std::string& Table::column_type(const ColumnID column_id) const { return _column_types.at(column_id); }
 
-std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) { return _chunks.at(chunk_id); }
+std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) {
+  auto guard = std::lock_guard<std::mutex>(_chunk_access_mutex);
+  return _chunks.at(chunk_id);
+}
 
-std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const { return _chunks.at(chunk_id); }
+std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
+  auto guard = std::lock_guard<std::mutex>(_chunk_access_mutex);
+  return _chunks.at(chunk_id);
+}
 
 void Table::compress_chunk(const ChunkID chunk_id) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  auto thread_pool = std::vector<std::thread>{_column_types.size()};
+  const auto& raw_chunk = get_chunk(chunk_id);  // get_chunk performs range check, so we are safe
+  auto compressed_chunk = std::make_shared<Chunk>();
+  auto compressed_segments = std::vector<std::shared_ptr<AbstractSegment>>{_column_types.size()};
+  auto segments_vec_mutex = std::mutex{};
+
+  auto compress_segment = [&raw_chunk, &compressed_segments, &segments_vec_mutex](const auto segment_type,
+                                                                                  const auto column_index) {
+    resolve_data_type(segment_type, [&](const auto data_type_t) {
+      // figure out the type of the segment
+      using ColumnDataType = typename decltype(data_type_t)::type;
+      auto guard = std::lock_guard<std::mutex>(segments_vec_mutex);
+      // add compressed segment to the map of segment (column) name to segment to later add to chunk in correct order
+      compressed_segments[column_index] =
+          std::make_shared<DictionarySegment<ColumnDataType>>(raw_chunk->get_segment(column_index));
+    });
+  };
+
+  // create one worker per segment, they will start running the moment they are created
+  for (auto column_index = ColumnID{0}; column_index < _column_types.size(); ++column_index) {
+    thread_pool[column_index] = std::thread(compress_segment, _column_types[column_index], column_index);
+  }
+
+  // wait for all threads to finish execution
+  for (auto&& thread : thread_pool) {
+    thread.join();
+  }
+
+  // add the compressed segments to the chunk in correct order
+  for (auto column_index = ColumnID{0}; column_index < _column_types.size(); ++column_index) {
+    compressed_chunk->add_segment(compressed_segments[column_index]);
+  }
+
+  auto guard = std::lock_guard<std::mutex>(_chunk_access_mutex);
+  // replace existing chunk with the new, compressed one
+  _chunks[chunk_id] = compressed_chunk;
+  /*
+   * We currently believe the following is the case:
```

Comment:
Good comment that cleared up a few things for us. I would move it to the top though.